sql/api/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -1382,6 +1382,21 @@ class Column(val node: ColumnNode) extends Logging {
    */
   def over(): Column = over(Window.spec)
 
+  /**
+   * Marks this column reference as an outer reference for subqueries.
+   *
+   * @group subquery
+   * @since 4.0.0
+   */
+  def outer(): Column = withOrigin {
+    node match {
+      case attr: UnresolvedAttribute if !attr.isMetadataColumn =>
+        Column(LazyOuterReference(attr.nameParts, attr.planId))
+      case _ =>
+        throw new IllegalArgumentException(
```

Comment:
we should throw an internal error.

Reply:
This is a user-facing API, so I feel throwing an internal error is weird. I'm ok to change the exception type to other user-facing exception, though.